src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -3107,5 +3103,233 @@ .locals init (char V_0)
   IL_006d:  ret
 }");
         }
+
+        [Fact]
+        public void New9PatternsSemanticModel_01()
+        {
+            // Tests for the semantic model in new patterns as of C# 9.0.
+            var source =
+@"
+using System;
+class Program
+{
+    void M(object o)
+    {
+        const int N = 12;
+        const char A = 'A';
+        const char Z = 'Z';
+        const char a = 'a';
+        const char z = 'z';
+        const char c0 = '0';
+        const char c9 = '9';
+        switch (o)
+        {
+            // Parenthesized patterns
+            case ((N, N)): break;
+            case (((long), (long))): break;
+            case ((N)): break;
+            // type patterns
+            case (int, int): break;
+            case (System.Int64, System.Int32): break;
+            case int: break;
+            // Conjunctive and disjunctive patterns
+            case (>= A and <= Z) or (>= a and <= z): break;
+            // Negated patterns
+            case not (> c0 and < c9): break;
+        }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source, options: TestOptions.DebugDll);
+            compilation.VerifyDiagnostics(
+                );
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+
+            var patterns = tree.GetRoot().DescendantNodes().OfType<PatternSyntax>().ToArray();
+            Assert.Equal(31, patterns.Length);
+            for (int i = 0; i < 31; i++)
+            {
+                var pattern = patterns[i];
+                AssertEmpty(model.GetSymbolInfo(pattern));
+                switch (i)
+                {
+                    case 0:
+                        Assert.Equal("((N, N))", pattern.ToString());
+                        Assert.Equal(SyntaxKind.ParenthesizedPattern, pattern.Kind());
+                        Assert.Equal("System.Object", model.GetTypeInfo(pattern).Type.ToTestDisplayString());
```

Comment:
I'll admit this surprises me. Why is it object?

Reply:
And I guess that follows for the rest of the `Object` assertions below.

Reply:
In this way patterns are the exact opposite of expressions.

In an expression, types flow from the inside to the outside, possibly getting converted as they go.

In a pattern, types flow from the outside to the inside, possibly getting converted as they go.

In this particular case, the expression coming in is of type `object`.  However, it is converted to the type `ITuple` by the pattern in order to be matched.

---
In reply to: [391898883](https://github.com/dotnet/roslyn/pull/42357#discussion_r391898883) [](ancestors = 391898883)